File: CppPractice/Project1/Project1/Leetcode142.cpp
#include <string>

#include <vector>

#include <iostream>

#include <stdlib.h>

#include <time.h>

#include <sstream>

#include <algorithm>

#include <map>

#include <bitset>

using namespace std;


struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode* detectCycle(ListNode* head) {
        vector<ListNode*> record;
        ListNode* cur = head;
        while (cur)
        {
            if (find(record.begin(),record.end(),cur)!=record.end())
            {
                return cur;
            }
            record.push_back(cur);
            cur = cur->next;

        }
        return NULL;

    }
};

File: Pythonleetcode/Leetcode142.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return None
        a=head
        exist_node_list=[]
        while a!=None:
            if a in exist_node_list:
                return a
            exist_node_list.append(a)
            a = a.next
        return None

